core/src/test/java/com/linecorp/armeria/client/endpoint/healthcheck/PartialHealthCheckStrategyTest.java

```diff
@@ -0,0 +1,279 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client.endpoint.healthcheck;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import com.linecorp.armeria.client.Endpoint;
+
+public class PartialHealthCheckStrategyTest {
+    private static final double HEALTHY = 1;
+    private static final double UNHEALTHY = 0;
+
+    private static final double MAX_RATIO = 0.9;
+    private static final int MAX_COUNT = 5;
+    private PartialHealthCheckStrategy maxRatioStrategy;
+    private PartialHealthCheckStrategy maxCountStrategy;
+    private List<Endpoint> candidatesForMaxRatio;
+    private List<Endpoint> candidatesForMaxCount;
+
+    private static List<Endpoint> createCandidates(int size) {
+        final Random random = new Random();
+
+        return IntStream.range(0, size)
+                        .mapToObj(i -> Endpoint.of("dummy" + random.nextInt()))
+                        .collect(Collectors.toList());
+    }
+
+    private static void assertCandidates(List<Endpoint> act, List<Endpoint> exp) {
+        assertThat(act).hasSize(exp.size());
+
+        for (Endpoint expCandidate : exp) {
+            assertThat(act.contains(expCandidate)).isTrue();
+        }
+    }
+
+    @BeforeEach
+    void beforeEach() {
+        maxRatioStrategy = new PartialHealthCheckStrategyBuilder().maxEndpointRatio(MAX_RATIO).build();
+        maxCountStrategy = new PartialHealthCheckStrategyBuilder().maxEndpointCount(MAX_COUNT).build();
+
+        candidatesForMaxRatio = createCandidates(10);
+        candidatesForMaxCount = createCandidates(6);
+
+        maxRatioStrategy.updateCandidates(candidatesForMaxRatio);
+        maxCountStrategy.updateCandidates(candidatesForMaxCount);
+    }
+
+    @Test
+    void getCandidatesWhenBeforeFirstUpdateCandidates() {
+        maxRatioStrategy = new PartialHealthCheckStrategyBuilder().maxEndpointRatio(MAX_RATIO)
+                                                                  .build();
+
+        assertThat(maxRatioStrategy.getCandidates()).isEmpty();
+    }
+
+    @Test
+    void getCandidatesWhenUpdateCandidatesByEmpty() {
+        maxRatioStrategy.updateCandidates(new ArrayList<>());
+        assertThat(maxRatioStrategy.getCandidates()).isEmpty();
+    }
+
+    @Test
+    void getCandidates() {
+        maxRatioStrategy.updateCandidates(candidatesForMaxRatio);
+
+        final List<Endpoint> actCandidates = maxRatioStrategy.getCandidates();
+        assertThat(actCandidates).hasSize(9);
+
+        actCandidates.forEach(
+                actCandidate -> assertThat(candidatesForMaxRatio.contains(actCandidate)).isTrue());
```

Comment:
Could you also check if the `selectedCandidates` contains distinct endpoints? i.e. no duplicate

Reply:
add test code